# Used Cars on eBay Kleinanzeigen #

For this project we will work with a dataset fromeBay Kleinanzeigen, a classified section of the German eBAy website.

The dataset is actually a 50.000 data points sample from a bigger one, in order for our code to run smoothly on Jupyter.

We will be focusing on cleaning the data of this Dataset and analyze it in a second time. 

Let's using Numpy and Pandas to help us.

First we need to import our csv file and let's take a look at what's inside.

In [1]:
import numpy as np
import pandas as pd

autos = pd.read_csv("autos.csv", encoding = "Windows-1252")
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

Our DataFrame has 20 columns and 50 000 rows. Here are a few quick observations we can make by looking at the first few rows.
- There seems to be some values missing in a few columns, such as "vehicleType" or "gearbox". 
- Some of the columns like the price are a mix of numbers and letters, and so are not considered as "Int" but object (just like in "odometer")
- In the name of the column and in the data there are Upper and Lower cases mixed, we will first focus on the column name to help us work on this dataset more easily.


In [2]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']

autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

We changed a few columns' names :

- yearOfRegistration to registration_year
- monthOfRegistration to registration_month
- notRepairedDamage to unrepaired_damage
- dateCreated to ad_created

And modify the rest of the names to get rid of all the uppercases. 

Let's now take a look at the dataset a bit more in details to get to know what we should work on in priority


In [3]:
autos.describe

<bound method DataFrame.describe of               date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/

There are some columns with mostly one values and that we can drop such as :  "offer_type", "nr_of_pictures","seller"

There are columns that need to be investigated : "ad_created","model","name","unrepaired_damage" 

There are columns with numeric datas stored as text : "price","odometer" => Let's deal with that.

In [4]:
autos["price"] = autos["price"].str.replace("$","").str.replace(",","").astype(float)
autos.rename(columns={"price" : "price_dollar"},inplace = True)
autos["odometer"] = autos["odometer"].str.replace(",","").str.replace("km","").astype(int)
autos.rename(columns={"odometer" : "odometer_km"},inplace = True)

We may have clean the "price_dollar" and "odometer_km" columns in terms of syntax but let's now look at the values, to see if there are no unrealistic values.

We are going to print :
- the number of unique values
- the information about the mean, the max, the min...
- the 5 biggest and lowest values with their number of occurence in the column

In [5]:
print(autos["price_dollar"].unique().shape)
print(autos["price_dollar"].describe())
print(autos["price_dollar"].value_counts().sort_index(ascending= True).head())
print(autos["price_dollar"].value_counts().sort_index(ascending= False).head())

(2357,)
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_dollar, dtype: float64
0.0    1421
1.0     156
2.0       3
3.0       1
5.0       2
Name: price_dollar, dtype: int64
99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
Name: price_dollar, dtype: int64


In [6]:
#Let's take a look at the most expensives cars in the dataset

print(autos[autos["price_dollar"] > 999999])

              date_crawled                                               name  \
2897   2016-03-12 21:50:57   Escort_MK_1_Hundeknochen_zum_umbauen_auf_RS_2000   
7814   2016-04-04 11:53:31                                        Ferrari_F40   
11137  2016-03-29 23:52:57  suche_maserati_3200_gt_Zustand_unwichtig_laufe...   
22947  2016-03-22 12:54:19                         Bmw_530d_zum_ausschlachten   
24384  2016-03-21 13:57:51                            Schlachte_Golf_3_gt_tdi   
27371  2016-03-09 15:45:47                                         Fiat_Punto   
39377  2016-03-08 23:53:51                        Tausche_volvo_v40_gegen_van   
39705  2016-03-22 14:58:27                        Tausch_gegen_gleichwertiges   
42221  2016-03-08 20:39:05                                  Leasinguebernahme   
47598  2016-03-31 18:56:54  Opel_Vectra_B_1_6i_16V_Facelift_Tuning_Showcar...   
47634  2016-04-04 21:25:21                                        Ferrari_FXX   

       seller offer_type  p

We decided to get rid of all the cars that are more expensive than the Ferrari_FXX, which is worth 3.890.000 dollars and the cars that cost less than 1$ (as Ebay is a biding website, it is realistic that a lot of persons put their car at 1$ to begin with the bids)

In [7]:
autos = autos[autos["price_dollar"].between(1,3890001)]

We still have more than 47.000 entries in our dataset after getting rid of the outliers. Now let's see the "odometer_km" values.

In [8]:
print(autos["odometer_km"].unique().shape)
print(autos["odometer_km"].describe())
print(autos["odometer_km"].value_counts().sort_index(ascending= True).head())
print(autos["odometer_km"].value_counts().sort_index(ascending= False).head())

(13,)
count     48571.000000
mean     125767.536184
std       39791.893858
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64
5000     837
10000    253
20000    762
30000    780
40000    815
Name: odometer_km, dtype: int64
150000    31417
125000     5058
100000     2115
90000      1734
80000      1415
Name: odometer_km, dtype: int64


The values are gathered by range, which is not very accurate as most of the cars are in the +150.000 kms category.

### The dates ###

Now let's take care of the columns where dates appear. There are five of them : 
- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website

The first three are considered as string by Pandas and the last two are considered as numeric values.
Let's determine what is the time range for our dataset.

In [9]:
print(autos["date_crawled"].value_counts(normalize=True, dropna=False).sort_index().head(5))
print(autos["date_crawled"].value_counts(normalize=True, dropna=False).sort_index().tail(5))



2016-03-05 14:06:30    0.000021
2016-03-05 14:06:40    0.000021
2016-03-05 14:07:04    0.000021
2016-03-05 14:07:08    0.000021
2016-03-05 14:07:21    0.000021
Name: date_crawled, dtype: float64
2016-04-07 14:30:09    0.000021
2016-04-07 14:30:26    0.000021
2016-04-07 14:36:44    0.000021
2016-04-07 14:36:55    0.000021
2016-04-07 14:36:56    0.000021
Name: date_crawled, dtype: float64


For the "date_crawled" column we can see that the dates go from the 5th of March 2016 to the 7th of April 2016.

In [10]:
print(autos["ad_created"].value_counts(normalize=True, dropna=False).sort_index().head(5))
print(autos["ad_created"].value_counts(normalize=True, dropna=False).sort_index().tail(5))


2015-06-11 00:00:00    0.000021
2015-08-10 00:00:00    0.000021
2015-09-09 00:00:00    0.000021
2015-11-10 00:00:00    0.000021
2015-12-05 00:00:00    0.000021
Name: ad_created, dtype: float64
2016-04-03 00:00:00    0.038850
2016-04-04 00:00:00    0.036894
2016-04-05 00:00:00    0.011818
2016-04-06 00:00:00    0.003253
2016-04-07 00:00:00    0.001256
Name: ad_created, dtype: float64


For the "ad_created" one it goes from the 11th of June 2014 to the 7th of April 2016

In [11]:
print(autos["last_seen"].value_counts(normalize=True, dropna=False).sort_index().head(5))
print(autos["last_seen"].value_counts(normalize=True, dropna=False).sort_index().tail(5))

2016-03-05 14:45:46    0.000021
2016-03-05 14:46:02    0.000021
2016-03-05 14:49:34    0.000021
2016-03-05 15:16:11    0.000021
2016-03-05 15:16:47    0.000021
Name: last_seen, dtype: float64
2016-04-07 14:58:44    0.000062
2016-04-07 14:58:45    0.000021
2016-04-07 14:58:46    0.000021
2016-04-07 14:58:48    0.000062
2016-04-07 14:58:50    0.000062
Name: last_seen, dtype: float64


Finally when it comes to the "last_seen" column, we can see that the dates range from the 5th of March 2016 to the 07th of April 2016.

We now know that we have a dataset representing all the visits on the Ebay Kleinanzeigen website from the 05/03/2016 (EU Date) to the 07/04/2016 (EU Date). During this timelapse visitors on the website could access to a broad range of car, that were posted between the 11/06/2014 and the 07/04/2016.

By taking a look at the "registration_year" column we can see how old are those cars

In [12]:
autos["registration_year"].describe()

count    48571.000000
mean      2004.754257
std         88.638548
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

We can see than approximatively one quarter of them are older than 17 year, half of them have more than 12 years and one quarter of them are less than 8 years old.

However those numbers are impacted by some false values : the min and max of this series are obviously out of range (1000 and 9999). The maximum cannot be above 2016 as the max of "ad_created" is in 2016. For the minimum we will play it safe and assume that no car from before 1900 could be on the website.

In [17]:
autos = autos[autos['registration_year'].between(1900, 2016)]
autos['registration_year'].value_counts(normalize = True).head(10)


2000    0.067599
2005    0.062887
1999    0.062073
2004    0.057896
2003    0.057811
2006    0.057211
2001    0.056461
2002    0.053248
1998    0.050614
2007    0.048772
Name: registration_year, dtype: float64

It seems like the majority of the cars listed are less than 20 years old, which makes sense.

### Brands ### 

Now let's examine how the brand of the car is affecting the prices.
For that we will use aggregation on the 20 main brands represented on the website.

In [34]:
autos["brand"].value_counts(ascending = False).head(10)

volkswagen       9864
bmw              5138
opel             5022
mercedes_benz    4503
audi             4041
ford             3264
renault          2201
peugeot          1393
fiat             1197
seat              853
Name: brand, dtype: int64

In [41]:
brand_10_price = {}
top10 = autos["brand"].value_counts(ascending = False).head(10).index
for i in top10:
    brand1 = autos[autos["brand"] == i]
    mean_price = brand1["price_dollar"].mean()
    brand_10_price[i] = int(mean_price)
    
print(brand_10_price)
   

{'renault': 2474, 'seat': 4397, 'ford': 4054, 'opel': 2975, 'mercedes_benz': 8628, 'audi': 9336, 'fiat': 2813, 'peugeot': 3094, 'bmw': 8571, 'volkswagen': 5604}


We can see that the most expensive cars are from the brand "Audi", followed by "Mercedes Benz" and "BMW". Which makes sense as they are pretty expensive cars. 

Now let's focus on the 6 first brand and try to see if there is a correlation between the mean price and the average mileage for those cars. First let's calculate the average mileage for each of those brands.


In [42]:
brand_6_mileage = {}
top6 = autos["brand"].value_counts(ascending = False).head(6).index
for i in top6:
    brand1 = autos[autos["brand"] == i]
    mean_mileage = brand1["odometer_km"].mean()
    brand_6_mileage[i] = int(mean_mileage)
    
print(brand_6_mileage)
   

{'ford': 124266, 'opel': 129310, 'mercedes_benz': 130788, 'audi': 129157, 'bmw': 132575, 'volkswagen': 128711}


Let's use the Pandas Series and Dataframe constructor to compare the mean_mileage and the mean_price

In [43]:
mean_price = pd.Series(brand_10_price)